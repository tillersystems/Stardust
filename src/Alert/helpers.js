import { css } from 'styled-components';

/**
 * Get Alert Position
 *
 * @param {string} position // Position of the alert on the screen.
 *
 */

export function getAlertPosition(position) {
  return {
    'top-left': css`
      left: 0;
      top: 0;
    `,
    'top-center': css`
      left: 50%;
      top: 0;
    `,
    'top-right': css`
      right: 0;
      top: 0;
    `,
    'bottom-left': css`
      left: 0;
      bottom: 0;
    `,
    'bottom-center': css`
      left: 50%;
      bottom: 0;
    `,
    'bottom-right': css`
      right: 0;
      bottom: 0;
    `,
  }[position];
}

/**
 * Get Position Animation
 *
 * @param {string} position // Position of the alert on the screen.
 *
 */
export function getPositionAnimation(position) {
  return {
    'top-left': {
      x: 0,
      y: -60,
    },
    'top-center': {
      x: '-50%',
      y: -60,
    },
    'top-right': {
      x: 0,
      y: -60,
    },
    'bottom-left': {
      x: 0,
      y: 60,
    },
    'bottom-center': {
      x: '-50%',
      y: 60,
    },
    'bottom-right': {
      x: 0,
      y: 60,
    },
  }[position];
}
