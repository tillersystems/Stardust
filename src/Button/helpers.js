import { css } from 'styled-components';
import { darken } from 'polished';

/**
 * getAppearance
 *
 * Helper function to render the right css
 * compared to the appearance given to it.
 *
 * @param {object} palette // The color's pallete of the theme.
 * @param {string} appearance // The button appearance.
 */

export const getAppearance = ({ palette }, appearance) => {
  return {
    default: css`
      color: ${palette.darkBlue};
      background: none;
      svg > path {
        fill: ${palette.darkBlue};
      }
    `,
    primary: css`
      background: ${palette.primary.default}
        linear-gradient(
          0deg,
          ${({ theme: { palette } }) => palette.whiteOpacity(0)} 0%,
          ${({ theme: { palette } }) => palette.whiteOpacity(0.1)} 100%
        );
      border: 1px solid ${palette.primary.dark};
      box-shadow: inset 0 0.2rem 0 0 ${palette.whiteOpacity(0.1)};

      /* Hovered  */
      &:hover:not([disabled]):not(:active) {
        background: ${palette.primary.default};
        box-shadow: none;
      }

      /* Activated  */
      &:active {
        background-color: ${palette.primary.dark};
        border: 1px solid ${palette.primary.darker};
        box-shadow: inset 0 0 0 0.1rem ${palette.primary.darker};
      }
    `,
    secondary: css`
      color: ${palette.spaceGrey};
      background: linear-gradient(180deg, ${palette.white} 0%, ${palette.mysticGrey} 100%);
      border: 1px solid ${palette.lightGrey};
      svg > path {
        fill: ${palette.spaceGrey};
      }
    `,
    success: css`
      background: ${palette.success.default}
        linear-gradient(0deg, ${palette.whiteOpacity(0)} 0%, ${palette.whiteOpacity(0.1)} 100%);
      border: 1px solid ${palette.success.dark};
      box-shadow: inset 0 0.2rem 0 0 ${palette.whiteOpacity(0.1)};

      /* Hovered  */
      &:hover:not([disabled]):not(:active) {
        background: ${palette.success.default};
        box-shadow: none;
      }

      /* Activated  */
      &:active {
        background-color: ${palette.success.dark};
        border: 1px solid ${palette.success.darker};
        box-shadow: inset 0 0 0 0.1rem ${palette.success.darker};
      }
    `,
    failure: css`
      background: ${palette.failure.default}
        linear-gradient(0deg, ${palette.whiteOpacity(0)} 0%, ${palette.whiteOpacity(0.1)} 100%);
      border: 1px solid ${palette.failure.dark};
      box-shadow: inset 0 0.2rem 0 0 ${palette.whiteOpacity(0.1)};

      /* Hovered  */
      &:hover:not([disabled]):not(:active) {
        background: ${palette.failure.default};
        box-shadow: none;
      }

      /* Activated  */
      &:active {
        background-color: ${palette.failure.dark};
        border: 1px solid ${palette.failure.darker};
        box-shadow: inset 0 0 0 0.1rem ${palette.failure.darker};
      }
    `,
    google: css`
      background: ${palette.googleBrandRed};
      border: 0.1rem solid ${palette.googleBrandRed};
      box-shadow: inset 0 0.2rem 0 0 ${palette.whiteOpacity(0.1)};

      /* Hovered  */
      &:hover:not([disabled]):not(:active) {
        background: ${darken(0.1, palette.googleBrandRed)};
      }
    `,
  }[appearance];
};

/**
 * getSize
 *
 * Helper function to render the right css
 * compared to the size given to him.
 *
 * @param {object} fonts // The font theme.
 * @param {string} size // The button size.
 */

export const getSize = ({ fonts }, size, icon) => {
  return {
    small: css`
      padding: ${!icon ? '0.35rem 1.4rem' : '0.35rem 1.6rem 0.35rem 0.9rem'};
    `,
    default: css`
      padding: ${!icon ? '0.75rem 2.2rem' : '0.75rem 2.4rem 0.75rem 1.4rem'};
    `,
    large: css`
      padding: ${!icon ? '0.7rem 2.4rem' : '0.7rem 2.4rem 0.7rem 1.8rem'};
      font-size: ${fonts.size.big};
      line-height: 1.5;
    `,
  }[size];
};
