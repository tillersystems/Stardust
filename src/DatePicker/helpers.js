import { DateTime, Interval } from 'luxon';

const CALENDAR_WEEKS = 6; // Weeks displayed on calendar

/**
 * Get previous month
 * Gets the month and year before the given month and year
 *
 * @param {luxon.DateTime} date - A date value
 *
 * @return {luxon.DateTime}
 */
export const getPreviousMonth = date => date.minus({ month: 1 });

/**
 * Get next month
 * Gets the month and year after the given month and year
 *
 * @param {luxon.DateTime} date - A date value
 *
 * @return {luxon.DateTime}
 */
export const getNextMonth = date => date.plus({ month: 1 });

/**
 * Is same day
 * Checks if two date values are the same day
 *
 * @param {luxon.DateTime} date - A date value
 * @param {luxon.DateTime} currentDate - A date value
 *
 * @return {boolean}
 */
export const isSameDay = (date, currentDate) => date.hasSame(currentDate, 'day');

/**
 * Is same month
 * Checks if two date values are the in the same month
 *
 * @param {luxon.DateTime} date - A date value
 * @param {luxon.DateTime} currentDate - A date value
 *
 * @return {boolean}
 */
export const isSameMonth = (date, currentDate) => date.hasSame(currentDate, 'month');

/**
 * Is in next month
 * Whether the selected date is in the next month w.r.t. the whole calendar
 *
 * @param {luxon.DateTime} date - A date value
 * @param {luxon.DateTime} currentDate - A date value
 *
 * @return {boolean}
 */
export const isInNextMonth = (date, currentDate) => {
  const monthStart = date.startOf('month');
  return monthStart.hasSame(currentDate.plus({ month: 1 }), 'month');
};

/**
 * Is in pervious month
 * Whether the selected date is in the pervious month
 *
 * @param {luxon.DateTime} date - A date value
 * @param {luxon.DateTime} currentDate - A date value
 *
 * @return {boolean}
 */
export const isInPrevMonth = (date, currentDate) => {
  const monthEnd = date.endOf('month');
  return monthEnd.hasSame(currentDate.minus({ month: 1 }), 'month');
};

/**
 * Is interval
 * Check if an object is an Interval
 *
 * @param {object} value
 *
 * @return {boolean}
 */
export const isInterval = value => Interval.isInterval(value);

/**
 * Calendar builder for a month in the specified year
 * Returns an array of the calendar dates.
 * Each calendar date is represented as an object => {YYYY, MM, DD}
 *
 * @param {luxon.DateTime} date - A date value
 *
 * @return {array}
 */
export const getCalendarDates = date => {
  const { daysInMonth, year, month } = date;

  // Get the month's first day
  const startOfMonth = date.startOf('month');
  const monthFirstDay = DateTime.fromISO(startOfMonth).weekday;

  // Get number of days to be displayed from previous and next months
  // These ensure a total of 42 days (6 weeks) displayed on the calendar
  const daysFromPrevMonth = monthFirstDay - 1;
  const daysFromNextMonth = CALENDAR_WEEKS * 7 - (daysFromPrevMonth + daysInMonth);

  // Get the previous and next months and years
  const { month: prevMonth, year: prevMonthYear } = getPreviousMonth(date);
  const { month: nextMonth, year: nextMonthYear } = getNextMonth(date);

  // Get number of days in previous month
  const prevMonthDays = date.minus({ month: 1 }).daysInMonth;

  // Builds dates to be displayed from previous month
  const prevMonthDates = [...new Array(daysFromPrevMonth)].map((n, index) => {
    const day = index + 1 + (prevMonthDays - daysFromPrevMonth);
    return { year: prevMonthYear, month: prevMonth, day };
  });

  // Builds dates to be displayed from current month
  const thisMonthDates = [...new Array(daysInMonth)].map((n, index) => {
    const day = index + 1;
    return { year, month, day };
  });

  // Builds dates to be displayed from next month
  const nextMonthDates = [...new Array(daysFromNextMonth)].map((n, index) => {
    const day = index + 1;
    return { year: nextMonthYear, month: nextMonth, day };
  });

  // Combines all dates from previous, current and next months
  return [...prevMonthDates, ...thisMonthDates, ...nextMonthDates];
};

/**
 * Returns date bounded to optional min or max
 *
 * @param {DateTime} date - the date to bound
 * @param {DateTime} minDate - minimum date
 * @param {DateTime} maxDate - maximum date
 *
 * @returns {DateTime}
 */
export const minMaxDate = (date, minDate, maxDate) => {
  const minBoundedDate = minDate ? DateTime.max(minDate, date) : date;
  const boundedDate = maxDate ? DateTime.min(maxDate, minBoundedDate) : minBoundedDate;

  return boundedDate;
};
