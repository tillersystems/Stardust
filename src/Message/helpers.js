export const getStatusBackgroundColor = ({ palette }, type) => {
  return {
    info: palette.primary.default,
    success: palette.success.default,
    error: palette.failure.default,
    warning: palette.warning.default,
  }[type];
};
