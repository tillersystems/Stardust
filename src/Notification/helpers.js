import { css } from 'styled-components';

/**
 * Get Alert Position
 *
 * @param {string} placement - Placement of the notification on the screen.
 *
 * @return {array}
 */

export function getNotificationPosition(padding, placement) {
  return {
    'top-left': css`
      left: ${padding};
      top: ${padding};
    `,
    'top-center': css`
      left: 50%;
      top: ${padding};
    `,
    'top-right': css`
      right: ${padding};
      top: ${padding};
    `,
    'bottom-left': css`
      left: ${padding};
      bottom: ${padding};
    `,
    'bottom-center': css`
      left: 50%;
      bottom: ${padding};
    `,
    'bottom-right': css`
      right: ${padding};
      bottom: ${padding};
    `,
  }[placement];
}

/**
 * Get Position Animation
 *
 * @param {string} placement - Placement of the notification on the screen.
 *
 * @return {object}
 */
export function getPositionAnimation(placement) {
  return {
    'top-left': {
      enter: {
        x: '0%',
        y: '0%',
      },
      exit: {
        x: '-100%',
        y: '0%',
      },
    },
    'top-center': {
      enter: {
        x: '-50%',
        y: '0%',
      },
      exit: {
        x: '-50%',
        y: '-100%',
      },
    },
    'top-right': {
      enter: {
        x: '0%',
        y: '0%',
      },
      exit: {
        x: '100%',
        y: '0%',
      },
    },
    'bottom-left': {
      enter: {
        x: '0%',
        y: '0%',
      },
      exit: {
        x: '-100%',
        y: '0%',
      },
    },
    'bottom-center': {
      enter: {
        x: '-50%',
        y: '0%',
      },
      exit: {
        x: '-50%',
        y: '100%',
      },
    },
    'bottom-right': {
      enter: {
        x: '0%',
        y: '0%',
      },
      exit: {
        x: '100%',
        y: '0%',
      },
    },
  }[placement];
}

/**
 * Timer
 *
 * @param {function} callback - Callback called when timer is done.
 * @param {number} delay - The delay of the timer.
 *
 * @return {jsx}
 */
export function Timer(callback, delay) {
  let timer = delay;
  let start = delay;
  let remaining = delay;

  this.clear = () => clearTimeout(timer);

  this.pause = () => {
    clearTimeout(timer);
    remaining -= Date.now() - start;
  };

  this.resume = () => {
    start = Date.now();
    clearTimeout(timer);
    timer = setTimeout(callback, remaining);
  };

  this.resume();
}
