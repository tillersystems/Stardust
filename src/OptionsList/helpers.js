import { memo } from 'react';

export const hasItemPropsAreEqual = (prev, next) => {
  return (
    prev.onChange === next.onChange &&
    prev.disabled === next.disabled &&
    prev.label === next.label &&
    prev.value === next.value &&
    prev.isSelected === next.isSelected
  );
};

export const memoItem = Component => memo(Component, hasItemPropsAreEqual);
