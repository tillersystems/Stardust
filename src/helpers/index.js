export { dark, light, getLightness, setColorDark, whiteOpacity } from './colors';
export { formatCompactedNumber, formatNumber, formatUnitNumber, UNIT_SYMBOLS } from './formatting';
