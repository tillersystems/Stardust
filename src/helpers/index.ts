export { dark, light, getLightness, setColorDark, whiteOpacity } from './colors';
export { formatCompactedNumber, formatNumber } from './formatting';
