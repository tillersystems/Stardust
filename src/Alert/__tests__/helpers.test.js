import { getAlertPosition, getPositionAnimation } from '../helpers';

describe('<Alert /> -- helpers', () => {
  describe('getAlertPosition', () => {
    it('should return Alert Css Position', () => {
      const expected = getAlertPosition('top-left');

      expect(expected).toMatchSnapshot();
    });
  });

  describe('getPositionAnimation', () => {
    it('should return Animation position', () => {
      const expected = getPositionAnimation('top-center');

      expect(expected).toMatchSnapshot();
    });
  });
});
