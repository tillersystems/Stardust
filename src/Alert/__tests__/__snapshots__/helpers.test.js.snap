// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert /> -- helpers getAlertPosition should return Alert Css Position 1`] = `
Array [
  "
      left: 0;
      top: 0;
    ",
]
`;

exports[`<Alert /> -- helpers getPositionAnimation should return Animation position 1`] = `
Object {
  "x": "-50%",
  "y": -60,
}
`;
